File: examples/search-algorithms/sut-binary-search.ts
/**
 * Binary Search SUT
 *
 * Iterative binary search on the pre-sorted data array.
 * Counts each mid-point comparison performed during the search.
 */

interface SearchInput {
	data: number[];
	sortedData: number[];
	target: number;
}

interface SearchResult {
	found: boolean;
	index: number;
	comparisons: number;
	executionTimeMs: number;
}

export function createSut() {
	return {
		id: "binary-search",
		config: {},
		run: (input: SearchInput): Promise<SearchResult> => {
			const { sortedData, target } = input;
			const start = performance.now();

			let comparisons = 0;
			let index = -1;
			let low = 0;
			let high = sortedData.length - 1;

			while (low <= high) {
				const mid = Math.floor((low + high) / 2);
				comparisons++;

				if (sortedData[mid] === target) {
					index = mid;
					break;
				} else if (sortedData[mid] < target) {
					low = mid + 1;
				} else {
					high = mid - 1;
				}
			}

			const executionTimeMs = performance.now() - start;

			return Promise.resolve({
				found: index !== -1,
				index,
				comparisons,
				executionTimeMs,
			});
		},
	};
}

File: examples/search-algorithms/sut_binary_search.py
"""Binary Search SUT — iterative binary search on sorted data."""

import time


def create_sut():
    def run(input):
        sorted_data = input["sortedData"]
        target = input["target"]
        start = time.perf_counter()

        comparisons = 0
        index = -1
        low = 0
        high = len(sorted_data) - 1

        while low <= high:
            mid = (low + high) // 2
            comparisons += 1

            if sorted_data[mid] == target:
                index = mid
                break
            elif sorted_data[mid] < target:
                low = mid + 1
            else:
                high = mid - 1

        execution_time_ms = (time.perf_counter() - start) * 1000

        return {
            "found": index != -1,
            "index": index,
            "comparisons": comparisons,
            "executionTimeMs": execution_time_ms,
        }

    return {"id": "binary-search", "config": {}, "run": run}
